code/modules/fabrication/fabricator_bioprinter.dm

```diff
@@ -9,28 +11,75 @@
 	base_storage_capacity = list(
 		/decl/material/solid/meat = SHEET_MATERIAL_AMOUNT * 100
 	)
-	var/datum/dna/loaded_dna_datum
+	var/datum/dna/loaded_dna //DNA for biological organs
+
+/obj/machinery/fabricator/bioprinter/get_nano_template()
+	return "fabricator_bioprinter.tmpl"
+
+/obj/machinery/fabricator/bioprinter/make_order(datum/fabricator_recipe/recipe, multiplier)
+	var/datum/fabricator_build_order/order = ..()
+	//Keep these in the order so changing settings while queueing things up won't screw up older orders in the queue
+	LAZYSET(order.data, "dna", loaded_dna)
+	return order
 
-/obj/machinery/fabricator/bioprinter/do_build(var/datum/fabricator_recipe/recipe, var/amount)
+/obj/machinery/fabricator/bioprinter/do_build(datum/fabricator_build_order/order)
 	. = ..()
+	//Fetch params as they were when the order was passed
+	var/datum/dna/D = LAZYACCESS(order.data, "dna")
 	for(var/obj/item/organ/O in .)
-		if(loaded_dna_datum)
-			O.set_dna(loaded_dna_datum)
+		if(D)
+			O.set_dna(D)
 		O.status |= ORGAN_CUT_AWAY
 
 /obj/machinery/fabricator/bioprinter/attackby(obj/item/W, mob/user)
 	if(istype(W,/obj/item/chems/syringe))
 		var/obj/item/chems/syringe/S = W
 		if(REAGENT_VOLUME(S.reagents, /decl/material/liquid/blood))
-			var/loaded_dna = REAGENT_DATA(S.reagents, /decl/material/liquid/blood)
-			if(islist(loaded_dna))
-				var/weakref/R = loaded_dna["donor"]
+			var/sample = REAGENT_DATA(S.reagents, /decl/material/liquid/blood)
+			if(islist(sample))
+				var/weakref/R = sample["donor"]
 				var/mob/living/carbon/human/H = R.resolve()
 				if(H && istype(H) && H.species && H.dna)
-					loaded_dna_datum = H.dna && H.dna.Clone()
+					loaded_dna = H.dna && H.dna.Clone()
 					to_chat(user, SPAN_INFO("You inject the blood sample into \the [src]."))
-					S.reagents.clear_reagents()
+					S.reagents.remove_any(BIOPRINTER_BLOOD_SAMPLE_SIZE) 
+					//Tell nano to do its job
+					SSnano.update_uis(src)
 					return TRUE
 		to_chat(user, SPAN_WARNING("\The [src] displays an error: no viable blood sample could be obtained from \the [W]."))
 		return TRUE
 	. = ..()
+
+/obj/machinery/fabricator/bioprinter/OnTopic(user, href_list, state)
+	if(href_list["flush_dna"])
+		if(fab_status_flags & FAB_BUSY)
+			state("Can't flush DNA while printing in progress!")
+		else
+			loaded_dna = null
+		. = TOPIC_REFRESH
+	. = ..()
```

Comment:
By not returning in the previous block, you will never set `TOPIC_REFRESH` and potentially make further calls. Generally it's simplest to return after this kind of block, unless sending multiple commands at once is necessary on the front end.

Reply:
True.. Probably should just run the parent proc before instead.. Because technically, href_list can have more than one entry per call.

Reply:
done